### ISSUE TITLE
publishToConfluence fails with error "Domain attribute "company.org" violates RFC 2109: domain must start with a dot"

### ISSUE THREAD

**Issue by hflocken** [docToolchain/docToolchain#273]
Hi,

we are trying to import our generated HTML pages into Confluence. I have tested this directly after OOP 2019 and it was working fine. But now I'm getting the error 'Domain attribute "company.org" violates RFC 2109: domain must start with a dot' when publishing to Confluence.

I would appreciate any hint what might be going wrong here.

```> Task :publishToConfluence
publish /Users/myuser/Data/source/myservices/build/docs/html5/myservices_org.html
arc42
Cookie rejected [crowd.p.token_key="eHs-qUkNJXcUK06_GRo3pQAAAAAACgABcDM0NDM3NA==", version:1, domain:company.org, path:/, expiry:null] Domain attribute "company.org" violates RFC 2109: domain must start with a dot
> created page null
1. Architecture
    image: ./images/portal_services_context_diagram.png
    image: ./images/building_blocks/connect_and_smart_mobility_services.png
    image: ./images/building_blocks/alert_services.jpg
    image: ./images/building_blocks/car_control.jpg
    image: ./images/building_blocks/car_finder.jpg
    image: ./images/building_blocks/charge_management_ilf.jpg
    image: ./images/building_blocks/climate.jpg
    image: ./images/building_blocks/e-control.jpg
    image: ./images/building_blocks/navigation_plus.jpg
    image: ./images/building_blocks/pre_heater.jpg
    image: ./images/building_blocks/trip_control.jpg
    image: ./images/building_blocks/vehicle_api.jpg
    image: ./images/building_blocks/vehicle_capability_service.jpg
    image: ./images/deployment/cloud_accounts_and_environments.jpeg
    image: ./images/deployment/cloud_production.jpg
    image: ./images/deployment/cloud_preproduction.jpg
    image: ./images/deployment/cloud_tools.jpg
    image: ./images/deployment/service_vehicle.jpg
    image: ./images/deployment/service_vcs.jpg
    image: ./images/deployment/charge_management.jpg
Cookie rejected [crowd.p.token_key="eHs-qUkNJXcUK06_GRo3pQAAAAAACgABcDM0NDM3NA==", version:1, domain:company.org, path:/, expiry:null] Domain attribute "company.org" violates RFC 2109: domain must start with a dot
> created page null
Cookie rejected [crowd.p.token_key="eHs-qUkNJXcUK06_GRo3pQAAAAAACgABcDM0NDM3NA==", version:1, domain:company.org, path:/, expiry:null] Domain attribute "company.org" violates RFC 2109: domain must start with a dot


FAILURE: Build failed with an exception.

* Where:
Script '/Users/myuser/Data/source/myservices/docToolchain/scripts/publishToConfluence.gradle' line: 24

* What went wrong:
Execution failed for task ':publishToConfluence'.
> groovyx.net.http.HttpResponseException: Not Found

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

* Get more help at https://help.gradle.org

BUILD FAILED in 4s
2 actionable tasks: 1 executed, 1 up-to-date```

**Comment by rdmueller** [comment]
thanx for reporting this issue.
The code for `publishToConfluence` hasn't been touched in a while and I don't recognize this exception from Jira.

It seems that the calling system tries to send a cookie which is not accepted by a receiving system.

The cookie shows my that you use Atlassian CROWD as authentication.

Could it be that you use a modified version of docToolchain? I think the master version does not support cookies.

**Comment by hflocken** [comment]
No, we are not using a modified version. As mentioned, I installed docToolchain around January 25th and it used to work then. My colleague installed the latest version just this week and she ran into exactly the same problem. I think I can remember that a Confluence update was performed some time ago. Is it possible that the import is no longer working with the new version of Confluence (we are using version 6.4.3)?

**Comment by rdmueller** [comment]
hm. I use the cloud instance of confluence and it is hard to tell which version is behind it.

you stacktrace mentions problems with cookies and I don't know where they are set. Do you have any idea?

**Comment by hflocken** [comment]
We use a hosted instance of confluence. Unfortunately I have no clue about the Confluence API and how it uses cookies.

**Comment by bodewig** [comment]
This might be  - and most likely is - the HTTP client talking about a Cookie set by the server. The `domain` of a Cookie is only set by the server. It could originate from a proxy in between doctoolchain and the server.

RFC2109 used to require the domain value to start with a dot. This has been obsoleted by [RFC 2965](https://tools.ietf.org/html/rfc2965) in October **2000** which drops the requirement. So whoever talks about the Cookie is rejecting it because of an ancient RFC that has been superseeded. I've been out of the loop for some time, but really have a hard time believing any HTTP client used by Groovy might be that old.

Any chance you could run this with `--debug` (and purge sensitive information)?

**Comment by bodewig** [comment]
Thanks, so it is `org.apache.http.client.protocol.ResponseProcessCookies` rejecting the Cookie. Strange, I'll take a look there.

The code is trying to read https://confluence.company.org/confluence/rest/api/content/null/child/attachment?filename=portal_services_context_diagram.png and gets a 404 - this might be related to a failed authentication because the necessary Cookie has been omitted. Can you access the URI when logged in? "null" sounds a bit strange on itself, so maybe the Cookie is a red herring.

The error happens while uploading attachments, does the page itself exist in Confluence when the task fails?

**Comment by hflocken** [comment]
No, the page doesn't exist in confluence. This is actually one of the picture I included in the asciiDoc document and that should be included in the HTML as well. As no upload has happened, this file cannot be present in Confluence from my point of view.

**Comment by bodewig** [comment]
The page itself should have been created when your log says

```
15:58:48.407 [QUIET] [system.out] > created page null
15:58:48.407 [QUIET] [system.out] 1. Architecture
15:58:48.419 [QUIET] [system.out]     image: ./images/services_context_diagram.png
```

The exception happens when the task tries to figure out whether the image already exists as an attachment to the page (it will only overwrite changed images so it needs to figure out whether the image exists first and if so whether it has changed second). It fails because the page doesn't seem to exist. Does it exist at all - maybe under a different name?

The `null` in `created page null` is supposed to be the id of the generated page, obtained from the response to the POST request that created the page. This one doesn't seem to contain the `id` where we expect it. This is the real problem as far as I can tell. Confluence accepts the POST with a 200 status code but its response doesn't look like one the code was expecting.

```
15:58:48.390 [DEBUG] [org.apache.http.impl.conn.DefaultClientConnection] Sending request: POST /confluence/rest/api/content HTTP/1.1
15:58:48.405 [DEBUG] [org.apache.http.impl.conn.DefaultClientConnection] Receiving response: HTTP/1.1 200 OK
```

I'm not really sure how to make progress here. Your server says things are fine but doesn't respond with an `id` in its JSON response. Maybe you could dump the `page` object or `page.data` before line 453 in `asciidoc2confluence.groovy`? There might be a hint of why the upload operation has failed.

The Cookie stuff is most likely not related at all. The code is using oldish versions of Groovy's HttpBuilder and Apache HttpClient but even those would be able to deal with Cookies setting a domain with a leading dot if configured properly. I don't think the current code would send back any cookies the server has tried to set in previous requests anyway, so this point seems to be moot and is only distracting us.

**Comment by hflocken** [comment]
I have added the following to Groovy:
```       //create a page
        trythis {
            page = api.post(contentType: ContentType.JSON,
                            requestContentType: ContentType.JSON,
                            path: 'content', body: request, headers: headers)
        }
        println "> created page "+page?.data?.id
        println "> DUMP1 "+page?.data
        println "> DUMP2 "+page
``` 

And I get the following output:
```Cookie rejected [crowd.p.token_key="mOTZBCVJNfjhQQaCTP0xwgAAAAAACgABcDM0NDM3NA==", version:1, domain:porsche.org, path:/, expiry:null] Domain attribute "porsche.org" violates RFC 2109: domain must start with a dot
> created page null
> DUMP1 null
> DUMP2 groovyx.net.http.HttpResponseDecorator@72d2a70e
```

This is all I was able to produce with my non-existing Groovy knowledge. Please let me know if I should change the debug out.

**Comment by bodewig** [comment]
This says `page.data` is `null` which should contain the pre-parsed response body of the request.

It's bee quite some time since I last looked at the API we use. We need to figure out what the server is actually sending back. If you know how to sniff your network traffic getting the raw response data might help. Alternatively we need to get there incrementally using what we have and will need to use use as a debugger. I'm really sorry about that but as we are currently not able to reproduce the error I'm not sure which other way of troubleshooting might work.

Could you please add

```
println "Content-Type: ${page.contentType}"
page.headers.each {
    println "${it.name} : ${it.value}"
}
println "Entity: ${page.entity}"
```

**Comment by hflocken** [comment]
Now I'm geeting the error 
```
> Response does not have a content-type header
```

**Comment by bodewig** [comment]
which is even more strange. Please remove the first `println` line above.

**Comment by hflocken** [comment]
This is what I get now:
```
Cookie rejected [crowd.p.token_key="mOTZBCVJNfjhQQaCTP0xwgAAAAAACgABcDM0NDM3NA==", version:1, domain:company.org, path:/, expiry:null] Domain attribute "company.org" violates RFC 2109: domain must start with a dot
> created page null
Header: Date : Thu, 21 Feb 2019 11:23:27 GMT
Header: Server : Apache-Coyote/1.1
Header: X-ASEN : SEN-2173426
Header: Content-Length : 0
Header: Keep-Alive : timeout=60, max=999
Header: Connection : Keep-Alive
Entity: org.apache.http.conn.BasicManagedEntity@450e1eeb
1. XX Architecture
Cookie rejected [crowd.p.token_key="mOTZBCVJNfjhQQaCTP0xwgAAAAAACgABcDM0NDM3NA==", version:1, domain:company.org, path:/, expiry:null] Domain attribute "company.org" violates RFC 2109: domain must start with a dot
> created page null
Header: Date : Thu, 21 Feb 2019 11:23:28 GMT
Header: Server : Apache-Coyote/1.1
Header: X-ASEN : SEN-2173426
Header: Content-Length : 0
Header: Keep-Alive : timeout=60, max=999
Header: Connection : Keep-Alive
Entity: org.apache.http.conn.BasicManagedEntity@5fa391f0
```

**Comment by bodewig** [comment]
`Content-Length : 0` doesn't look good.

We could try to read the entity's content, something like

```
page.entity.writeTo(System.out)
```

but would most likely end up either getting an exception or an empty output. So it is probably not worth the effort,

For some reason your server responds to the POST request with a 200 status code but no body at all. This is not exactly what https://developer.atlassian.com/cloud/confluence/rest/#api-content-post promises.

If this was a permission problem I'd expect a 401 or 403 response, could there be anybody in between that mapped 40x responses to 200 (I've seen this with "helpful" reverse proxies in the past). Is there any chance you can get hold of Confluence's logs to see what it has said about the request?

**Comment by rdmueller** [comment]
thanx for all this analysis output.

let's see what we've got:

- the response seems to be a 200 or at least no error status (have we validated the 200?)
- the response body is empty. A json with a page ID would be expected
- where does the "cookie rejected" message come from?
- the cookie is from the atlassian CROWD SSO service. I don't know where it will be set. docToolchain uses the basic auth mechanism for auth. I am even not sure if the used client would support cookies.

What I currently guess is that some kind of auth proxy or redirect adds the error message without changing the status code.

Another question is - what has changed that the export now suddenly fails.

@hflocken have you already asked your confluence team if they have changed something?

**Comment by bodewig** [comment]
Answering the questions that I can answer

200 is validated, see:

```
15:58:48.390 [DEBUG] [org.apache.http.impl.conn.DefaultClientConnection] Sending request: POST /confluence/rest/api/content HTTP/1.1
15:58:48.405 [DEBUG] [org.apache.http.impl.conn.DefaultClientConnection] Receiving response: HTTP/1.1 200 OK
```

"cookie rejected" comes deep from within Apache HttpClient which is used by `RestClient`. I'm not entirely sure about the exact HttpClient version but the message comes from this class (or a predecessor/ancestor of it): https://github.com/apache/httpcomponents-client/blob/4.4.x/httpclient/src/main/java/org/apache/http/impl/cookie/RFC2109DomainHandler.java#L87 - there are handlers for newer RFCs and I assume we could tell RestClient to use HttpClient in a more lenient setup.

I also suspect the client we used will not preserve cookies between requests.

**Comment by hflocken** [comment]
@bodewig & @rdmueller I'm still trying to get in contact with one of our Confluence admins...

**Comment by rdmueller** [comment]
@bodewig : the error message comes from *which* client? I doubt that it is the docToolchain client (there is no set-cookie header). So it would be a system (CROWD SSO?) in between....

if I could spent more time I would like to check out crowd...

**Comment by hflocken** [comment]
Would it be possible to add some wire logging to the Groovy script so that we can see what exactly is going on? That might also help us in finding the log entries in Confluence.

**Comment by bodewig** [comment]
@rdmueller the log message is from the HttpClient library used by the RestClient used from inside the groovy script. It is logged when the HttpClient parses the response. So yes, the Cookie is set by Confluence/Crowd/whoever the script is actually talking to.

https://hc.apache.org/httpcomponents-client-ga/logging.html says we could use "wire logging", not sure how Apache Commons Logging used by Apache HttpClient, Groovy and gradle interact, though.

**Comment by hflocken** [comment]
@rdmueller I had some mail exchange with our Confluence admins and I have now received their final statement that they are not planning to change anything. 

Below you can find the response. Unfortunately (for this thread) it is only in German, but I know you will understand it 😄 Basically they are saying that the Apache HTTP client might have been configured too strictly.

=====================================================================

Confluence / JIRA ist erreichbar unter confluence/Jira.company.org und verwenden crowd unter dem FQDN crowd.company.org. --> die Cookies lauten auf die Domain companies.org und nicht eben auf den Host.

Crowd erlaubt zwar die eingabe von ".company.org", speichert diese aber nicht. Dies vollkommen zu Recht, denn laut RFC 6265 soll Cleintseitig wie folgt verfahren werden:

https://tools.ietf.org/html/rfc6265#section-5.2.3

If the first character of the attribute-value string is %x2E ("."):

Let cookie-domain be the attribute-value without the leading %x2E
(".") character.

Otherwise:
Let cookie-domain be the entire attribute-value.

siehe auch Atlassian Ticket https://jira.atlassian.com/browse/CWD-5141

--> die RFC 2109 ist reichlich veraltet. Wie auch in https://github.com/docToolchain/docToolchain/issues/273#issuecomment-465604079 erwähnt 

--> Wir können und werden dies auch nicht beheben, denn es ist aus meiner Sicht eine zu strikte Client-Implementierung

Wenn der apache http client die Ursache ist - wie auch immer dieser in docToolChain integriert ist - so ist dieser ggf. zu strict initialisert worden. siehe u.a.

https://stackoverflow.com/questions/42003021/cookie-rejected-for-rest-api-call

und https://hc.apache.org/httpcomponents-client-4.2.x/tutorial/html/statemgmt.html

**Comment by rdmueller** [comment]
so, basically, CROWD sets a cookie which docToolchain rejects because of a too strict implementation.

**Comment by rdmueller** [comment]
closed by mistake

**Comment by bodewig** [comment]
I'm afraid things are not that easy, but maybe I'm, just too pessimistic. Even if we manage to get a proper configuration for Apache HttpClient through the various layers of indirection Groovy's RestClient put between our script and the HTTP Call the best thing we will achieve is the error message will be dropped. I'd be surprised if a Cookie sent be the first request was sent back by subsequent requests unless we change more things.

BTW, Apache HttpClient should be recent enough as it is as support for non-deprecated RFCs has been in there for years.

Right now the only authentication method supported by `publishToConfluence` is basic auth. If I understand https://confluence.atlassian.com/kb/basic-authentication-troubleshooting-guide-719095277.html correctly, this simply is not an option for Crowd at all and we'd have to add support for the specific flavor of OAuth2 used by Crowd in order to get through.

**Comment by rdmueller** [comment]
see also https://github.com/confluence-publisher/confluence-publisher/pull/154/files and https://github.com/confluence-publisher/confluence-publisher/issues/153
I think this is the same problem
